### PR TITLE
FEATURE: MSF-16455 support yellow WARNING status into RubySDK

### PR DESCRIPTION
**Pull request by tqtu**

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/207/ : SUCCESS in 8s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/208/ : SUCCESS in 6s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/209/ : SUCCESS in 8s

File: lib/gooddata/bricks/middleware/execution_result_middleware.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    end
+
+    class ExecutionResultMiddleware < Bricks::Middleware
+      def call(params)
+        params = self.class.update_execution_result(params, params["EXECUTION_RESULT_STATUS"], params["EXECUTION_RESULT_MESSAGE"])
+        @app.call(params)
+      end
+
+      def self.update_execution_result(params, status, message = "")
+        if status != ExecutionStatus::OK && status != ExecutionStatus::ERROR && status != ExecutionStatus::ERROR
```
Comment (phong-nguyen-duy) [inline lib/gooddata/bricks/middleware/execution_result_middleware.rb]: There are 2 ExecutionStatus::ERROR

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/210/ : SUCCESS in 4s

File: bin/run_brick.rb [hunk 2]
```diff
@@ -78,7 +78,7 @@ def handle_error(params, log, brick_type, error, error_message)
 
   log.info "action=#{brick_type}_execution status=start commit_hash=#{commit_hash} execution_id=#{execution_id}"
   brick_pipeline.call(params)
-  update_execution_result(execution_result_log_params, log, brick_type, ExecutionStatus::OK, '')
+  update_execution_result(execution_result_log_params, log, brick_type, ExecutionStatus::OK, '') unless params['EXECUTION_RESULT_COMPLETED']
```
Comment (danh-ung) [inline bin/run_brick.rb]: we should remove method update_execution_result in run_brick.rb, switch to use ExecutionResultMiddleware

File: lib/gooddata/bricks/middleware/execution_result_middleware.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      WARNING = 'WARNING'
+    end
+
+    class ExecutionResultMiddleware < Bricks::Middleware
+      def call(params)
+        params = self.class.update_execution_result(params, params["EXECUTION_RESULT_STATUS"], params["EXECUTION_RESULT_MESSAGE"])
+        @app.call(params)
+      end
+
+      def self.update_execution_result(params, status, message = "")
```
Comment (danh-ung) [inline lib/gooddata/bricks/middleware/execution_result_middleware.rb]: add ruby doc: Update process execution result when the script is executed inside a GDC ETL process execution (ruby bricks). Ruby bricks should update execution result at the end of script with status=WARNING or OK and summary message if the script exits normally. If any fatal error, script should update execution result with status=ERROR and error message, then throw exception to notify GDC platform that the script exists unexpectedly.

File: lib/gooddata/bricks/middleware/execution_result_middleware.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            message: message
+          }
+        }
+        update_result(params, result)
+      end
+
+      private
+
+      def self.update_result(params, result)
+        execution_result_logger_file = params['GDC_EXECUTION_RESULT_LOG_PATH'] || ENV['GDC_EXECUTION_RESULT_LOG_PATH']
```
Comment (danh-ung) [inline lib/gooddata/bricks/middleware/execution_result_middleware.rb]: Should have default value "#{log_directory}/#{execution_id}_result.json"

File: lib/gooddata/bricks/middleware/execution_result_middleware.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            status: status,
+            message: message
+          }
+        }
+        update_result(params, result)
+      end
+
+      private
+
+      def self.update_result(params, result)
```
Comment (danh-ung) [inline lib/gooddata/bricks/middleware/execution_result_middleware.rb]: can we set this is private?

File: lib/gooddata/bricks/middleware/execution_result_middleware.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        execution_result_logger_file = params['GDC_EXECUTION_RESULT_LOG_PATH'] || ENV['GDC_EXECUTION_RESULT_LOG_PATH']
+
+        return if execution_result_logger_file.nil?
+
+        File.open(execution_result_logger_file, 'w') { |file| file.write(JSON.pretty_generate(result)) }
+
+        params.merge('EXECUTION_RESULT_COMPLETED' => true)
```
Comment (danh-ung) [inline lib/gooddata/bricks/middleware/execution_result_middleware.rb]: no need to have a flag, just update it to file. If the method is called many time, last time will be recorded.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/211/ : SUCCESS in 5s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/212/ : SUCCESS in 5s

File: lib/gooddata/bricks/middleware/execution_result_middleware.rb (hunk 3 above)
Comment (tqtu) [reply lib/gooddata/bricks/middleware/execution_result_middleware.rb]: I updated the ruby doc.

File: lib/gooddata/bricks/middleware/execution_result_middleware.rb (hunk 4 above)
Comment (tqtu) [reply lib/gooddata/bricks/middleware/execution_result_middleware.rb]: I updated. The GDC_EXECUTION_RESULT_LOG_PATH is higher priority.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/217/ : SUCCESS in 5s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/218/ : SUCCESS in 5s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/221/ : SUCCESS in 7s

File: lib/gooddata/bricks/middleware/execution_result_middleware.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        update_result(params, result)
+      end
+
+      private
+
+      def self.update_result(params, result)
+        log_directory = params['GDC_LOG_DIRECTORY']
+        execution_id = params['GDC_EXECUTION_ID']
+        execution_result_logger_file = params['GDC_EXECUTION_RESULT_LOG_PATH'] || ENV['GDC_EXECUTION_RESULT_LOG_PATH']
+        execution_result_logger_file = execution_result_logger_file.nil? ? "#{log_directory}/#{execution_id}_result.json" : execution_result_logger_file
```
Comment (phong-nguyen-duy) [inline lib/gooddata/bricks/middleware/execution_result_middleware.rb]: log_directory and execution_id are passed by k8s executor manager only, so please check nil for backward compatibility
it should be: 
`execution_result_logger_file = "#{log_directory}/#{execution_id}_result.json" if execution_result_logger_file.nil? && !log_directory.nil?`

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/224/ : SUCCESS in 5s

File: lib/gooddata/bricks/middleware/execution_result_middleware.rb (hunk 7 above)
Comment (tqtu) [reply lib/gooddata/bricks/middleware/execution_result_middleware.rb]: Yes, I fixed

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/225/ : SUCCESS in 15s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/226/ : SUCCESS in 6s

File: lib/gooddata/bricks/middleware/execution_result_middleware.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    class ExecutionResultMiddleware < Bricks::Middleware
+      def call(params)
+        self.class.update_execution_result(params, params['EXECUTION_RESULT_STATUS'], params['EXECUTION_RESULT_MESSAGE'])
+        @app.call(params)
+      end
+
+      # Update process execution result when the script is executed inside a GDC ETL process execution (ruby bricks).
+      # Ruby bricks should update execution result at the end of script with status=WARNING or OK and summary message if the script exits normally.
+      # If any fatal error, script should update execution result with status=ERROR and error message,
+      # then throw exception to notify GDC platform that the script exists unexpectedly.
```
Comment (danh-ung) [inline lib/gooddata/bricks/middleware/execution_result_middleware.rb]: the script exists unexpectedly --> the script terminated unexpectedly

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/246/ : SUCCESS in 5s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/247/ : SUCCESS in 7s

File: lib/gooddata/bricks/middleware/execution_result_middleware.rb [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+      end
+
+      # Update process execution result when the script is executed inside a GDC ETL process execution (ruby bricks).
+      # Ruby bricks should update execution result at the end of script with status=WARNING or OK and summary message if the script exits normally.
+      # If any fatal error, script should update execution result with status=ERROR and error message,
+      # then throw exception to notify GDC platform that the script terminated unexpectedly.
+      # @param [HashMap] params contains GDC_EXECUTION_RESULT_LOG_PATH or GDC_LOG_DIRECTORY, GDC_EXECUTION_ID
+      # @param [ExecutionStatus] status execution status
+      # @param [String] message execution message
+      def self.update_execution_result(status = ExecutionStatus::OK, message = "")
```
Comment (danh-ung) [inline lib/gooddata/bricks/middleware/execution_result_middleware.rb]: Please remove status default value, we should call update_execution_result(ExecutionStatus::OK) than only update_execution_result.

File: lib/gooddata/lcm/lcm2.rb [hunk 10]
```diff
@@ -365,6 +365,8 @@ def perform(mode, params = {})
           end
         end
 
+        GoodData::Bricks::ExecutionResultMiddleware.update_execution_result
```
Comment (danh-ung) [inline lib/gooddata/lcm/lcm2.rb]: should called with explicit status: update_execution_result(GoodData::Bricks::ExecutionStatus::OK)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/248/ : SUCCESS in 8s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/249/ : SUCCESS in 11s

File: lib/gooddata/lcm/lcm2.rb (hunk 10 above)
Comment (tqtu) [reply lib/gooddata/lcm/lcm2.rb]: Yes, I updated all comment.
Now , I will reopen release ruby PR branch to verify again

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/251/ : SUCCESS in 7s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/252/ : SUCCESS in 7s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/254/ : SUCCESS in 5s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-merge-compliance-check/419/ : SUCCESS in 20s

**Pull request merged**